### PR TITLE
Improved synchronous Request/Response

### PR DESCRIPTION
**Pull request by johnsimons**
### Description
Added ability to do Request/Response through the `IBus` with a more clean and friendly API.

Here is what the API looks like:
```c#
SendContext<MyResponse> sendContext = bus.SynchronousRequestResponse<MyResponse>(new MyRequest(), new SynchronousOptions());
//or
SendContext<MyResponse> sendContext = bus.SynchronousRequestResponse<MyResponse>(new MyRequest(), new SynchronousLocalOptions());

MyResponse response = await sendContext.ResponseTask;
//or alternative
sendContext.ResponseTask.ContinueWith(task =>
{
     MyResponse response = task.Result;
     // ...
});
```

### Backwards compatibility
Backwards compatibility is being handled with the introduction of a *special message* called `LegacyEnumResponse<T>`.
So to use this new message, all we need to do is upgrade code like:
```c#
Bus.Return(MyEnum.Success);
```
to
```c#
Bus.Reply(new LegacyEnumResponse<MyEnum>(MyEnum.Success));
```

and on the client side, upgrade code like:
```c#
Bus.Send(new MyRequest()).Register<MyEnum>(result => { ... });
```
to
```c#
var response = bus.SynchronousRequestResponse<LegacyEnumResponse<MyEnum>>(new MyRequest(), new SynchronousOptions());
var legacyEnumResponse = await response.ResponseTask;
MyEnum result = legacyEnumResponse.Status;
```

### Summary of changes
- [x] Deprecate `ICallback`, as this is not needed
- [x] Deprecate `Sends` that return `ICallback`
- [x] Deprecate all public buddy classes that are part of old callback support
- [x] Introduce new `SynchronousRequestResponse` methods
- [x] Deprecate `bus.Return`
- [x] Ensure we can still use old style callbacks against new style
- [x] Create new nuget package

### Remarks
I've currently added this to the Core solution for convenience, but obviously can be pulled out into its own solution.

### Issues addressed
- Concerns raised in #2596
- #2390

### CODE REVIEW

**Comment by johnsimons** [comment]
@Particular/nservicebus do you think we should support this ?

**Comment by SzymonPobiega** [comment]
I like it

File: src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        /// <param name="requestMessage"></param>
+        /// <returns></returns>
+        public static SendContext<TResponse> RequestResponse<TResponse>(this IBus bus, object requestMessage)
+        {
+            Guard.AgainstNull(requestMessage, "requestMessage");
+            Guard.AgainstNull(bus, "bus");
+
+            var correlationId = Guid.NewGuid().ToString();
+
+            bus.SetMessageHeader(requestMessage, "NServiceBus.HasCallback", Boolean.TrueString);
```
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: I don't like 'HasCallback' name but that's an impl detail. I think we should figure out something more meaningful for the receiver.

Comment (johnsimons) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: Yeah, @andreasohlund mentioned in #2596 that he want a header, I let him decide what to call it :wink:

Comment (SzymonPobiega) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: `NServiceBus.Andreas` ? ;-)

Comment (johnsimons) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: I think is more `NServiceBus.ThisHeaderIsToSupportAFutureVersionOfRoutingThatWeHaventReleasedYet`

Comment (andreasohlund) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: `NServiceBus.CallbackRegistered` ?

**Comment by andreasohlund** [comment]
> I've currently added this to the NServiceBus.Core, but obviously can be pulled out into its own assembly.

How about we make this at least a separate asm in that pull to make sure we don't run into issues with internals? (repo split can wait)

**Comment by johnsimons** [comment]
> How about we make this at least a separate asm in that pull to make sure we don't run into issues with internals?

Sure, what should I call it ?

File: src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: Since RequestResponse is a pattern that is used when users do
bus.Send/bus.Reply (async) I propose we rename this to
`bus.SyncronousRequestResponse`
to make clear that there is a difference?

Comment (andreasohlund) [inline src/NServiceBus.Core/Callbacks/RequestResponseExtensions.cs]: >Sure, what should I call it ?

I suck at names but is `NServiceBus.Callbacks` to vague?

**Comment by danielmarbach** [comment]
Is the ServiceProvider interface really needed. I think we should add this abstraction

**Comment by johnsimons** [comment]
@danielmarbach are you saying we should expose it on the `IBus` interface ?

**Comment by johnsimons** [comment]
@Particular/nservicebus 
So we're now at the stage where we need to decide whether we are deprecating `bus.Return` and `ICallback` from the the core in v6 or only in v7 ?

**Comment by andreasohlund** [comment]
Since we've already started and v6 isn't out I'd say we do it now

**Comment by SeanFeldman** [comment]
+1 for v6

**Comment by johnsimons** [comment]
@andreasohlund you have added [this AT](https://github.com/Particular/NServiceBus/commit/f44eec6a21d9516f9ccc7b09d73e82d2a55ce158) that covers callbacks against local endpoint, I have removed that as part of these changes, see https://github.com/Particular/NServiceBus/commit/026a4db0b1c24136c75c49c2743a5021d053fec5, Do we still need it ?

**Comment by johnsimons** [comment]
@Particular/nservicebus I can't figure out a way to cover [this](https://github.com/Particular/NServiceBus/commit/97678ea4ee7731ad55b3707fabd9b1ea570f7ead) and [this](https://github.com/Particular/NServiceBus/commit/74543bf2fe05e4bdda3ba0ead061ea5d02bce6dd) AT using the new synchronous req/resp API, does anyone have any ideas apart from covering it with wire compat tests ?

**Comment by johnsimons** [comment]
@andreasohlund it seems we need to run this lot of AT against all our transports, see https://github.com/Particular/NServiceBus/commit/ae83ba83896c681b63dfb18fd135eddb2f6f053b, how do we handle permutations ?

**Comment by andreasohlund** [comment]
>you have added this AT that covers callbacks against local endpoint, I have removed that as part of these changes, see 026a4db, Do we still need it ?

I still think it's a valid test?


>it seems we need to run this lot of AT against all our transports, see ae83ba8, how do we handle permutations ?

We have a similar situation for other things as well (like a focused test suite for persitences running for all persisters  cc @SzymonPobiega ). Should we get on a call with @Particular/nservicebus and try to come up with a plan for how to handle AT that's split from the core but needs to run for all transports?

**Comment by johnsimons** [comment]
> Should we get on a call with @Particular/nservicebus and try to come up with a plan for how to handle AT that's split from the core

:+1:

**Comment by johnsimons** [comment]
> you have added this AT that covers callbacks against local endpoint, I have removed that as part of these changes, see 026a4db, Do we still need it ?
> I still think it's a valid test?

@andreasohlund can you expand, where would you see synchronous req/resp be used against the local endpoint ?

**Comment by johnsimons** [comment]
@andreasohlund I need to know because that influences the proposed API, so far we only have:
```c#
SendContext<MyResponse> sendContext = bus.SynchronousRequestResponse<MyResponse>(new MyRequest());
```
do we need a:
```c#
SendContext<MyResponse> sendContext = bus.SynchronousRequestResponse<MyResponse>(new MyRequest(), destination);
//and
SendContext<MyResponse> sendContext = bus.LocalSynchronousRequestResponse<MyResponse>(new MyRequest());
```

??

**Comment by danielmarbach** [comment]
@johnsimons @andreasohlund agree about splitting. We discussed it already. It would require some hard trickery around duck typing I think

**Comment by andreasohlund** [comment]
> LocalSynchronousRequestResponse

I'm actually still not happy with separating out the local sends a separate b.SendLocal(Object, SendLocalOptions). In my mind they are identical except for how we route the message and is just causing us grief. This would be a non issue if I could convince you guys to drop that.

I guess I'm proposing:

```

bus.Send(new MyMessage(),new SendOptions())
bus.Send(new MyMessage(),new SendOptions(Route.UsingStaticMessageMappings)) //same as above

bus.Send(new MyMessage(),new SendOptions(Route.ToLocalEndpoint)); //send local

bus.Send(new MyMessage(), new SendOptions(Route.ToNativeAddress("sales@{sqlserver connection string}")));

```

and then drop the special send local.

What do you think? cc @Particular/nservicebus

**Comment by danielmarbach** [comment]
Is Route an enum, a static class with Function delegates or ?

**Comment by johnsimons** [comment]
@andreasohlund I dislike the magic `Route` class, it feels weird that we need `Route` to specify an address, it feels like another `Address` class, sorry!

**Comment by johnsimons** [comment]
It doesn't look like we can allow a user to customise the messageid or/and correlationid, is this a problem @andreasohlund ?

**Comment by johnsimons** [comment]
Actually we can support users to set a custom correlationid, but not sure if we really need to, thoughts ?

**Comment by johnsimons** [comment]
@Particular/engineers 
This is now open to be nitpicked, just be aware, there are still a few more things to do, see list in description.

**Comment by andreasohlund** [comment]
>It doesn't look like we can allow a user to customise the messageid or/and correlationid, is this a problem

Why can't we do that?

Isn't it strange that we support managing the message id (for dedup reasons) when sending a message but not when sending a message that happen to have a callback?

**Comment by johnsimons** [comment]
Forgot to send an update, this is now all working fine, a user can
customise both of those

**Comment by johnsimons** [comment]
@Particular/engineers this is now complete, ready for reviewing.
Any questions let me know.

File: src/NServiceBus.Callbacks/CallbacksSupport.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿namespace NServiceBus.Features
+{
+    using System;
+    using NServiceBus.Callbacks;
+
+    class CallbacksSupport : Feature
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/CallbacksSupport.cs]: Rename to `Callbacks` or `CallbackSupport` ?

File: src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    using NServiceBus.Pipeline;
+    using NServiceBus.Pipeline.Contexts;
+
+    class ConvertLegacyEnumResponseToLegacyControlMessageBehavior : PhysicalOutgoingContextStageBehavior
+    {
+        public override void Invoke(Context context, Action next)
+        {
+            var instance = context.Get<object>("RequestResponse.OutgoingLogicalMessageInstance");
+
+            if (!CallbacksSupport.isLegacyEnumResponse(instance.GetType()))
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs]: `context.MessageType` should be available on the context? (https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Pipeline/Contexts/PhysicalOutgoingContext.cs#L37)

It was added recently, rebase?

File: src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+            next();
+        }
+
+        public class Registration : RegisterStep
+        {
+            public Registration()
+                : base("RequestResponse_convert_legacy_enum_response_to_control_enum_message", typeof(ConvertLegacyEnumResponseToLegacyControlMessageBehavior), "Converts the legacy response message to a control message to support backward compatibiliy")
+            {
+                InsertAfterIfExists("LogOutgoingMessage");
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs]: Didn't we conclude that the log behaviour would need to know about control messages so this shouldn't be needed?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        /// Sends a <paramref name="requestMessage"/> to the configured destination and returns back a <see cref="SendContext{TResponse}"/> that allows the use user to wait on.
+        /// </summary>
+        /// <typeparam name="TResponse">The response type.</typeparam>
+        /// <param name="bus">Object beeing extended.</param>
+        /// <param name="requestMessage">The request message.</param>
+        /// <param name="options">The options for the send.</param>
+        /// <returns>A synchronous request/response context.</returns>
+        public static SendContext<TResponse> SynchronousRequestResponse<TResponse>(this IBus bus, object requestMessage, SynchronousOptions options)
+        {
+            if (requestMessage == null)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Don't we have those `Guard.X` methods to do this?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+            var sendOptions = new SendOptions(options.Destination, options.CorrelationId);
+                
+            foreach (var header in options.Headers)
+            {
+                sendOptions.AddHeader(header.Key, header.Value);
+            }
+
+            sendOptions
+                .AddHeader("NServiceBus.HasCallback", Boolean.TrueString)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Since this is all about routing and we sort of agreed (might need to talk about that again) to start prefixing our "system" headers with $ I propose we use `$Routing.HasCallback` ?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                .SetCustomMessageId(options.MessageId);
+
+            var tcs = new TaskCompletionSource<TResponse>();
+
+            sendOptions.GetContext().Add("NServiceBus.RequestResponse.TCS", tcs);
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Heads up: After my `publish-options` pull this will need to be strongly typed:

```
sendOptions.GetContext().Set(new RequestResponseContext{ TaskCompletionSource = tcs })
```

cc @SzymonPobiega

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        /// Sends a <paramref name="requestMessage"/> to the configured destination and returns back a <see cref="SendContext{TResponse}"/> that allows the use user to wait on.
+        /// </summary>
+        /// <typeparam name="TResponse">The response type.</typeparam>
+        /// <param name="bus">Object beeing extended.</param>
+        /// <param name="requestMessage">The request message.</param>
+        /// <param name="options">The options for the send.</param>
+        /// <returns>A synchronous request/response context.</returns>
+        public static SendContext<TResponse> SynchronousRequestResponse<TResponse>(this IBus bus, object requestMessage, SynchronousLocalOptions options)
+        {
+            if (requestMessage == null)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Guards ?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+            var sendLocalOptions = new SendLocalOptions(options.CorrelationId);
+
+            foreach (var header in options.Headers)
+            {
+                sendLocalOptions.AddHeader(header.Key, header.Value);
+            }
+
+            sendLocalOptions
+                .AddHeader("NServiceBus.HasCallback", Boolean.TrueString)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Change header (see above)

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                .SetCustomMessageId(options.MessageId);
+
+            var tcs = new TaskCompletionSource<TResponse>();
+
+            sendLocalOptions.GetContext().Add("NServiceBus.RequestResponse.TCS", tcs);
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Heads up, strongly typed (see above)

File: src/NServiceBus.Callbacks/RequestResponseInvocationBehavior.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        public RequestResponseInvocationBehavior(RequestResponseMessageLookup requestResponseMessageLookup)
+        {
+            this.requestResponseMessageLookup = requestResponseMessageLookup;
+        }
+
+        public override void Invoke(Context context, Action next)
+        {
+            var invoked = HandleCorrelatedMessage(context.PhysicalMessage, context);
+
+            context.Set("NServiceBus.CallbackInvocation.CallbackWasInvoked", invoked);
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseInvocationBehavior.cs]: Use the new `context.ExtensionContext` to avoid the stringly typed boolean?

File: src/NServiceBus.Callbacks/RequestResponseMessageLookup.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+namespace NServiceBus.Callbacks
+{
+    using System.Collections.Concurrent;
+
+    class RequestResponseMessageLookup
+    {
+        readonly ConcurrentDictionary<string, object> messageIdToAsyncResultLookup = new ConcurrentDictionary<string, object>();
+
+        internal void RegisterResult(string messageId, object tcs)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseMessageLookup.cs]: Can be public since the class is internal? (nitpicking)

File: src/NServiceBus.Callbacks/SynchronousLocalOptions.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
+namespace NServiceBus
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    ///     Allows the users to control how the send is performed
+    /// </summary>
+    public class SynchronousLocalOptions
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/SynchronousLocalOptions.cs]: Think I've manage to sway @SimonCropp and @danielmarbach to skip the local option all together but I'll raise a separate pull for that later

File: src/NServiceBus.Callbacks/SynchronousOptions.cs [hunk 14]
```diff
@@ -0,0 +1,53 @@
+namespace NServiceBus
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    ///     Allows the users to control how the send is performed
+    /// </summary>
+    public class SynchronousOptions
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/SynchronousOptions.cs]: Just noting: With this we do not longer support callbacks for deferred messages.

Ie: bus.Defer(new MyMessage()).Register() //not supported anymore

File: src/NServiceBus.Callbacks/UpdateRequestResponseCorrelationTableBehavior.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        public UpdateRequestResponseCorrelationTableBehavior(RequestResponseMessageLookup lookup)
+        {
+            this.lookup = lookup;
+        }
+
+        public override void Invoke(Context context, Action next)
+        {
+            var sendOptions = context.DeliveryMessageOptions as SendMessageOptions;
+
+            if (sendOptions == null)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/UpdateRequestResponseCorrelationTableBehavior.cs]: With the `publish-options` pull in you can simplify this to a single

```
RequestResponseState state;
if(context.Extensions.TryGet(out state))
{
lookup.RegisterResult(context.MessageId, state.TaskCompletionSource);
}
 next();
```

File: src/NServiceBus.Core/Unicast/Behaviors/LoadHandlersConnector.cs [hunk 16]
```diff
@@ -20,9 +20,7 @@ public override void Invoke(LogicalMessageProcessingStageBehavior.Context contex
         {
             bool callbackInvoked;
 
-
-            // for now we cheat and pull it from the behavior context:
-            if (!context.TryGet(CallbackInvocationBehavior.CallbackInvokedKey, out callbackInvoked))
+            if (!context.TryGet("NServiceBus.CallbackInvocation.CallbackWasInvoked", out callbackInvoked))
             {
                 callbackInvoked = false;
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/Behaviors/LoadHandlersConnector.cs]: This one has always been vierd. I think we need a context.MessageWasProcessed flag that the callbacks behavior can set instead?

This way the LoadHandlersConnecor doesn't need to know about the callbacks?

File: src/NServiceBus.Core/Unicast/Behaviors/SerializeMessagesBehavior.cs [hunk 17]
```diff
@@ -22,12 +22,6 @@ public SerializeMessagesBehavior(IMessageSerializer messageSerializer,MessageMet
 
         public override void Invoke(OutgoingContext context, Action<PhysicalOutgoingContextStageBehavior.Context> next)
         {
-            if (context.IsControlMessage)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/Behaviors/SerializeMessagesBehavior.cs]: Awesome! * 10

File: src/NServiceBus.Core/Unicast/ContextualBus.cs [hunk 18]
```diff
@@ -263,32 +261,6 @@ public void Reply<T>(Action<T> messageConstructor)
             Send(messageConstructor, context);
         }
 
-        /// <summary>
-        /// <see cref="IBus.Return{T}"/>
-        /// </summary>
-        public void Return<T>(T errorCode)
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/ContextualBus.cs]: You won't be missed dear friend

**Comment by andreasohlund** [comment]
@johnsimons did a first pass, let me know if you want to get on a call so that I can clarify some of the comments?

Left to do after this pull:

* Move to separate repo
* Modify transports to skip skip the callback queue if this feature isnät enabled (rabbitmq + ssqlserver)
* Doco
Correct?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 7 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: :+1:

Comment (johnsimons) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: @andreasohlund can you send me a link to this change and reasons why ?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 6 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: YAGNI

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 5 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: No, `Guard` is internal to core.

File: src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs (hunk 3 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/ConvertLegacyEnumResponseToLegacyControlMessageBehavior.cs]: I don't see a need to change it.
From my perspective we may not need `context.MessageType` at all

File: src/NServiceBus.Callbacks/CallbacksSupport.cs (hunk 2 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/CallbacksSupport.cs]: Are u nitpicking ?
It is an internal class, not really following you naming strategy ?

File: src/NServiceBus.Core/Unicast/Behaviors/LoadHandlersConnector.cs (hunk 16 above)
Comment (johnsimons) [inline src/NServiceBus.Core/Unicast/Behaviors/LoadHandlersConnector.cs]: yeah, that could help solidify the contract between the core and other components, feel free to change it

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 10 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: I'm not sure we need strongly type in this scenario

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 8 above)
Comment (johnsimons) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: where ?

**Comment by johnsimons** [comment]
@andreasohlund thanks a lot for review, from what I gather there is nothing that needs changing.
If it is ok with you I'll merge this one before you proposed `publish-options` pull

File: src/NServiceBus.Callbacks/SynchronousOptions.cs (hunk 14 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/SynchronousOptions.cs]: That seems like a good thing

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: I don't see that header used anywhere. Am I missing something?

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: >YAGNI

Are you calling YAGNI on naming a header correctly? (remember that we have no good way to version headers so we'd have to live with it for ever)

>I don't see that header used anywhere. Am I missing something?

It will be used by the upcoming routing changes that will need to route messages with callbacks properly.

Could also be used in SI to highlight that messages had callbacks waiting for them

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 8 above)
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Strike that, Guard is internal to the core

File: src/NServiceBus.Callbacks/SynchronousOptions.cs (hunk 14 above)
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/SynchronousOptions.cs]: Agree, just noting it :)

File: src/NServiceBus.Callbacks/CallbacksSupport.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/CallbacksSupport.cs]: We need to use the string name in the sql and rabbit transport

```
if(Feature.IsEnabled("CallbackSupport "))
{
   EnableTheCallbackQueueReceiver();
} 
```

Not a spelling expert but isn't the`trailing `s` in CallbacksSupport wrong?

File: src/NServiceBus.Callbacks/RequestResponseExtensions.cs (hunk 6 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: Are we using a proper language here. I mean, is the fact that the message has a callback essential or is it any other property? I mean, can there be a different use case which should trigger the same routing behavior change but is not related to the callbacks. In such case maybe the more appropriate name would be `$Routing.ReplyToSpecificInstance`

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/RequestResponseExtensions.cs]: +1 I like that header better!

I would change to `$Routing.ReplyToSpecificEndpointInstance`

**Comment by andreasohlund** [comment]
@johnsimons think we're good to go here now. Please review my updates
